File: benchmarks/binary_tree/binary_tree.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int value;
    struct Node *left;
    struct Node *right;
} Node;

Node* new_node(int value) {
    Node *n = malloc(sizeof(Node));
    n->value = value;
    n->left = NULL;
    n->right = NULL;
    return n;
}

Node* insert(Node *root, int value) {
    if (root == NULL) {
        return new_node(value);
    }
    if (value < root->value) {
        root->left = insert(root->left, value);
    } else {
        root->right = insert(root->right, value);
    }
    return root;
}

long inorder_sum(Node *root) {
    if (root == NULL) return 0;
    return inorder_sum(root->left) + root->value + inorder_sum(root->right);
}

void free_tree(Node *root) {
    if (root == NULL) return;
    free_tree(root->left);
    free_tree(root->right);
    free(root);
}

int main(int argc, char *argv[]) {
    int n = argc > 1 ? atoi(argv[1]) : 100000;

    Node *root = NULL;

    // Insert values in a pattern that creates a balanced-ish tree
    // Using a simple LCG to generate pseudo-random values
    unsigned int seed = 12345;
    for (int i = 0; i < n; i++) {
        seed = seed * 1103515245 + 12345;
        int value = (seed >> 16) & 0x7FFF;
        root = insert(root, value);
    }

    // Sum all values via in-order traversal
    long sum = inorder_sum(root);
    printf("%ld\n", sum);

    free_tree(root);
    return 0;
}

File: benchmarks/binary_tree/binary_tree.js
class Node {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

function insert(root, value) {
    if (root === null) {
        return new Node(value);
    }
    if (value < root.value) {
        root.left = insert(root.left, value);
    } else {
        root.right = insert(root.right, value);
    }
    return root;
}

function inorderSum(root) {
    if (root === null) return 0;
    return inorderSum(root.left) + root.value + inorderSum(root.right);
}

const n = parseInt(process.argv[2]) || 100000;

const start = performance.now();

let root = null;

// Insert values using LCG for pseudo-random values
let seed = 12345n;
for (let i = 0; i < n; i++) {
    seed = (seed * 1103515245n + 12345n) % 4294967296n;
    const value = Number((seed >> 16n) & 0x7FFFn);
    root = insert(root, value);
}

// Sum all values via in-order traversal
const result = inorderSum(root);
const elapsed = performance.now() - start;
console.error(`TIME_MS:${elapsed.toFixed(2)}`);
console.log(result);
